### ISSUE TITLE
Consistent: org.openqa.selenium.WebDriverException: org.apache.http.conn.HttpHostConnectException: Connect to localhost:13956 [localhost/127.0.0.1, localhost/0:0:0:0:0:0:0:1] failed: Connection refused: connect

### ISSUE THREAD

**Issue by codingbest** [SeleniumHQ/selenium#4203]
## Meta -
OS:  Windows 7

Selenium Version:  3.3.1

Browser:  Chromedriver 

Browser Version:  Version 59.0.3071.104 (Official Build) (64-bit)

## Steps to reproduce -

```
org.openqa.selenium.WebDriverException: org.apache.http.conn.HttpHostConnectException: Connect to localhost:13956 [localhost/127.0.0.1, localhost/0:0:0:0:0:0:0:1] failed: Connection refused: connect
Build info: version: '3.3.1', revision: '5234b325d5', time: '2017-03-10 09:10:29 +0000'
System info: host: 'INCHEL-063', ip: '10.129.45.74', os.name: 'Windows 7', os.arch: 'amd64', os.version: '6.1', java.version: '1.8.0_131'
Driver info: driver.version: Driver
	at org.openqa.selenium.remote.service.DriverCommandExecutor.execute(DriverCommandExecutor.java:91)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:604)
	at org.openqa.selenium.remote.RemoteWebDriver.startSession(RemoteWebDriver.java:244)
	at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:131)
	at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:144)
	at org.openqa.selenium.chrome.ChromeDriver.<init>(ChromeDriver.java:178)
	at org.openqa.selenium.chrome.ChromeDriver.<init>(ChromeDriver.java:167)
	at org.openqa.selenium.chrome.ChromeDriver.<init>(ChromeDriver.java:124)
	at com.ma5.browserpick.Browser.InvokeDriver(Browser.java:26)
	at com.ma5.testcases.DataMigrationSuite.Browser(DataMigrationSuite.java:83)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at com.ma5.driver.Driver.readExcel(Driver.java:138)
	at com.ma5.driver.Driver.main(Driver.java:249)
Caused by: org.apache.http.conn.HttpHostConnectException: Connect to localhost:13956 [localhost/127.0.0.1, localhost/0:0:0:0:0:0:0:1] failed: Connection refused: connect
	at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:158)
	at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:353)
	at org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:380)
	at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:236)
	at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:184)
	at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:88)
	at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:184)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:71)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:55)
	at org.openqa.selenium.remote.internal.ApacheHttpClient.fallBackExecute(ApacheHttpClient.java:142)
	at org.openqa.selenium.remote.internal.ApacheHttpClient.execute(ApacheHttpClient.java:88)
	at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:296)
	at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:113)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:141)
	at org.openqa.selenium.remote.service.DriverCommandExecutor.execute(DriverCommandExecutor.java:82)
	... 15 more
Caused by: java.net.ConnectException: Connection refused: connect
	at java.net.DualStackPlainSocketImpl.waitForConnect(Native Method)
	at java.net.DualStackPlainSocketImpl.socketConnect(Unknown Source)
	at java.net.AbstractPlainSocketImpl.doConnect(Unknown Source)
	at java.net.AbstractPlainSocketImpl.connectToAddress(Unknown Source)
	at java.net.AbstractPlainSocketImpl.connect(Unknown Source)
	at java.net.PlainSocketImpl.connect(Unknown Source)
	at java.net.SocksSocketImpl.connect(Unknown Source)
	at java.net.Socket.connect(Unknown Source)
	at org.apache.http.conn.socket.PlainConnectionSocketFactory.connectSocket(PlainConnectionSocketFactory.java:74)
	at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:141)
	... 30 more
```

**Comment by p0deje** [comment]
Please provide [concise reproducible test case](http://sscce.org/) so that we could act on this issue.

See [CONTRIBUTING.md](https://github.com/SeleniumHQ/selenium/blob/master/CONTRIBUTING.md#issue-contributions)

**Comment by rogerdc** [comment]
I am also receiving this error when running multiple test cases. 

Setup: Using .NET bindings. Visual Studio 2017 Community. Windows 10. Java 1.8.0_131

Versions of packages from NuGet.org: 
Selenium.WebDriver: 3.4.0
Selenium.WebDriver.ChromeDriver: 2.30.0.1
NUnit: 3.7.1

Steps: 

1. Create two test cases that should both pass using NUnit tags that use ChromeDriver.
2. In the Visual Studio Test Explorer window, run both of the tests as part of the same run (eg, use the "Run All" button."

Expected result:
Both test cases pass.

Actual result: 
The first test case passes, but the second fails with a stack trace similar to: 

```
Test Name:	ViewAPIAppsNoneAssociated
Test FullName:	NWSAutomatedTests.APIVault.TestCases.APIVaultTests.ViewAPIAppsNoneAssociated
Test Source:	C:\Users\Roger.Cook\Code\NICServices\NWSAutomatedTests\NWSAutomatedTests\APIVault\TestCases\APIVaultTests.cs : line 72
Test Outcome:	Failed
Test Duration:	0:00:06.011

Result StackTrace:	
at OpenQA.Selenium.Remote.RemoteWebDriver.UnpackAndThrowOnError(Response errorResponse)
   at OpenQA.Selenium.Remote.RemoteWebDriver.Execute(String driverCommandToExecute, Dictionary`2 parameters)
   at OpenQA.Selenium.Remote.RemoteWebDriver.FindElement(String mechanism, String value)
   at OpenQA.Selenium.Remote.RemoteWebDriver.FindElementByCssSelector(String cssSelector)
   at OpenQA.Selenium.By.<>c__DisplayClass1e.<CssSelector>b__1c(ISearchContext context)
   at OpenQA.Selenium.By.FindElement(ISearchContext context)
   at OpenQA.Selenium.Remote.RemoteWebDriver.FindElement(By by)
   at OpenQA.Selenium.Support.PageObjects.DefaultElementLocator.LocateElement(IEnumerable`1 bys)
   at OpenQA.Selenium.Support.PageObjects.WebElementProxy.get_Element()
   at OpenQA.Selenium.Support.PageObjects.WebElementProxy.Invoke(IMessage msg)
   at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)
   at OpenQA.Selenium.IWebElement.get_Displayed()
   at OpenQA.Selenium.Support.UI.ExpectedConditions.<>c__DisplayClass3e.<ElementToBeClickable>b__3d(IWebDriver driver)
   at OpenQA.Selenium.Support.UI.DefaultWait`1.Until[TResult](Func`2 condition)
   at NWSAutomatedTests.Common.PageObjects.LoginPage.AzureActiveDirectoryLogin() in C:\Users\Roger.Cook\Code\NICServices\NWSAutomatedTests\NWSAutomatedTests\NWSAutomatedTests\PageObjects\LoginPage.cs:line 29
   at NWSAutomatedTests.APIVault.TestCases.APIVaultTests.APITestSetup() in C:\Users\Roger.Cook\Code\NICServices\NWSAutomatedTests\NWSAutomatedTests\APIVault\TestCases\APIVaultTests.cs:line 23
--TearDown
   at OpenQA.Selenium.Remote.RemoteWebDriver.UnpackAndThrowOnError(Response errorResponse)
   at OpenQA.Selenium.Remote.RemoteWebDriver.Execute(String driverCommandToExecute, Dictionary`2 parameters)
   at OpenQA.Selenium.Remote.RemoteWebDriver.FindElement(String mechanism, String value)
   at OpenQA.Selenium.Remote.RemoteWebDriver.FindElementByCssSelector(String cssSelector)
   at OpenQA.Selenium.By.<>c__DisplayClass1e.<CssSelector>b__1c(ISearchContext context)
   at OpenQA.Selenium.By.FindElement(ISearchContext context)
   at OpenQA.Selenium.Remote.RemoteWebDriver.FindElement(By by)
   at OpenQA.Selenium.Support.PageObjects.DefaultElementLocator.LocateElement(IEnumerable`1 bys)
   at OpenQA.Selenium.Support.PageObjects.WebElementProxy.get_Element()
   at OpenQA.Selenium.Support.PageObjects.WebElementProxy.Invoke(IMessage msg)
   at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)
   at OpenQA.Selenium.IWebElement.get_Displayed()
   at OpenQA.Selenium.Support.UI.ExpectedConditions.<>c__DisplayClass3e.<ElementToBeClickable>b__3d(IWebDriver driver)
   at OpenQA.Selenium.Support.UI.DefaultWait`1.Until[TResult](Func`2 condition)
   at NWSAutomatedTests.APIVault.PageObjects.APIVaultMenuBar.ClickLogout() in C:\Users\Roger.Cook\Code\NICServices\NWSAutomatedTests\NWSAutomatedTests\APIVault\PageObjects\APIVaultMenuBar.cs:line 47
   at NWSAutomatedTests.APIVault.TestCases.APIVaultTests.APITestTearDown() in C:\Users\Roger.Cook\Code\NICServices\NWSAutomatedTests\NWSAutomatedTests\APIVault\TestCases\APIVaultTests.cs:line 32
Result Message:	
OpenQA.Selenium.WebDriverException : Unexpected error. System.Net.WebException: Unable to connect to the remote server ---> System.Net.Sockets.SocketException: No connection could be made because the target machine actively refused it 127.0.0.1:63628
   at System.Net.Sockets.Socket.DoConnect(EndPoint endPointSnapshot, SocketAddress socketAddress)
   at System.Net.ServicePoint.ConnectSocketInternal(Boolean connectFailure, Socket s4, Socket s6, Socket& socket, IPAddress& address, ConnectSocketState state, IAsyncResult asyncResult, Exception& exception)
   --- End of inner exception stack trace ---
   at System.Net.HttpWebRequest.GetRequestStream(TransportContext& context)
   at System.Net.HttpWebRequest.GetRequestStream()
   at OpenQA.Selenium.Remote.HttpCommandExecutor.Execute(Command commandToExecute)
   at OpenQA.Selenium.Remote.DriverServiceCommandExecutor.Execute(Command commandToExecute)
   at OpenQA.Selenium.Remote.RemoteWebDriver.Execute(String driverCommandToExecute, Dictionary`2 parameters)
TearDown : OpenQA.Selenium.WebDriverException : Unexpected error. System.Net.WebException: Unable to connect to the remote server ---> System.Net.Sockets.SocketException: No connection could be made because the target machine actively refused it 127.0.0.1:63628
   at System.Net.Sockets.Socket.DoConnect(EndPoint endPointSnapshot, SocketAddress socketAddress)
   at System.Net.ServicePoint.ConnectSocketInternal(Boolean connectFailure, Socket s4, Socket s6, Socket& socket, IPAddress& address, ConnectSocketState state, IAsyncResult asyncResult, Exception& exception)
   --- End of inner exception stack trace ---
   at System.Net.HttpWebRequest.GetRequestStream(TransportContext& context)
   at System.Net.HttpWebRequest.GetRequestStream()
   at OpenQA.Selenium.Remote.HttpCommandExecutor.Execute(Command commandToExecute)
   at OpenQA.Selenium.Remote.DriverServiceCommandExecutor.Execute(Command commandToExecute)
   at OpenQA.Selenium.Remote.RemoteWebDriver.Execute(String driverCommandToExecute, Dictionary`2 parameters)
Result StandardOutput:	
Begin API Vault Test Setup
Begin API Vault Test Teardown

```

**Comment by rogerdc** [comment]
Note: This may not be specific to ChromeDriver. I am also getting similar failures and stack traces from Selenium.WebDriver.GeckoDriver.Win64 version 0.17.0.

For both Firefox and Chrome, if the test cases are run separately, they all pass. It's only when trying to run multiple test cases that the failure appears.

**Comment by rogerdc** [comment]
One other note to reproduce: Start up and shut down the browser in both test cases.

**Comment by cgoldberg** [comment]
> Start up and shut down the browser in both test cases.

why can't you post the code?

**Comment by rogerdc** [comment]
The code I originally reproduced this issue in is proprietary. Gotta feed the kids somehow. :)

I made a repository at https://github.com/rogerdc/SeleniumIssue4203Repro2.git that successfully reproduces the issue by searching for "Cheese" and "Fromage" at Google and Bing, respectively.

**Comment by cgoldberg** [comment]
are you sure these tests are executing sequentially and not in parallel?

**Comment by rogerdc** [comment]
They're executing sequentially according to what I see in the VS Test Explorer window. One goes to running state, then is marked passed before the other one goes to the running state.

**Comment by cgoldberg** [comment]
got a stack trace from this?  because the one you showed from your previous tests looks like execution is interleaved   so the first case runs it's TearDown, and closes all browser instances.. when the second case calls TearDown, the driver is already gone.

**Comment by rogerdc** [comment]
Test Name:	Test2
Test FullName:	SeleniumIssue4203Repro2.TestCases.SequentialTests.Test2
Test Source:	C:\Users\Tom\Source\Repos\selenium-issue-4203-repro2\SeleniumIssue4203Repro2\SeleniumIssue4203Repro2\TestCases\SequentialTests.cs : line 41
Test Outcome:	Failed
Test Duration:	0:00:08.165

Result StackTrace:	
at OpenQA.Selenium.Remote.RemoteWebDriver.UnpackAndThrowOnError(Response errorResponse)
   at OpenQA.Selenium.Remote.RemoteWebDriver.Execute(String driverCommandToExecute, Dictionary`2 parameters)
   at OpenQA.Selenium.Remote.RemoteWebDriver.FindElement(String mechanism, String value)
   at OpenQA.Selenium.Remote.RemoteWebDriver.FindElementById(String id)
   at OpenQA.Selenium.By.<>c__DisplayClass2.<Id>b__0(ISearchContext context)
   at OpenQA.Selenium.By.FindElement(ISearchContext context)
   at OpenQA.Selenium.Remote.RemoteWebDriver.FindElement(By by)
   at OpenQA.Selenium.Support.UI.ExpectedConditions.<>c__DisplayClass3b.<ElementToBeClickable>b__3a(IWebDriver driver)
   at OpenQA.Selenium.Support.UI.DefaultWait`1.Until[TResult](Func`2 condition)
   at SeleniumIssue4203Repro2.TestCases.SequentialTests.Test2() in C:\Users\Tom\Source\Repos\selenium-issue-4203-repro2\SeleniumIssue4203Repro2\SeleniumIssue4203Repro2\TestCases\SequentialTests.cs:line 43
--TearDown
   at OpenQA.Selenium.Remote.HttpCommandExecutor.CreateResponse(WebRequest request)
   at OpenQA.Selenium.Remote.HttpCommandExecutor.Execute(Command commandToExecute)
   at OpenQA.Selenium.Remote.DriverServiceCommandExecutor.Execute(Command commandToExecute)
   at OpenQA.Selenium.Remote.RemoteWebDriver.Execute(String driverCommandToExecute, Dictionary`2 parameters)
   at OpenQA.Selenium.Remote.RemoteWebDriver.Close()
   at SeleniumIssue4203Repro2.WrapperFactory.BrowserFactory.CloseAllDrivers() in C:\Users\Tom\Source\Repos\selenium-issue-4203-repro2\SeleniumIssue4203Repro2\SeleniumIssue4203Repro2\WrapperFactory\BrowserFactory.cs:line 90
   at SeleniumIssue4203Repro2.TestCases.SequentialTests.ShutdownDriver() in C:\Users\Tom\Source\Repos\selenium-issue-4203-repro2\SeleniumIssue4203Repro2\SeleniumIssue4203Repro2\TestCases\SequentialTests.cs:line 26
--WebException
   at System.Net.HttpWebRequest.GetResponse()
   at OpenQA.Selenium.Remote.HttpCommandExecutor.CreateResponse(WebRequest request)
--SocketException
   at System.Net.Sockets.Socket.DoConnect(EndPoint endPointSnapshot, SocketAddress socketAddress)
   at System.Net.ServicePoint.ConnectSocketInternal(Boolean connectFailure, Socket s4, Socket s6, Socket& socket, IPAddress& address, ConnectSocketState state, IAsyncResult asyncResult, Exception& exception)
Result Message:	
OpenQA.Selenium.WebDriverException : Unexpected error. System.Net.WebException: Unable to connect to the remote server ---> System.Net.Sockets.SocketException: No connection could be made because the target machine actively refused it 127.0.0.1:54785
   at System.Net.Sockets.Socket.DoConnect(EndPoint endPointSnapshot, SocketAddress socketAddress)
   at System.Net.ServicePoint.ConnectSocketInternal(Boolean connectFailure, Socket s4, Socket s6, Socket& socket, IPAddress& address, ConnectSocketState state, IAsyncResult asyncResult, Exception& exception)
   --- End of inner exception stack trace ---
   at System.Net.HttpWebRequest.GetRequestStream(TransportContext& context)
   at System.Net.HttpWebRequest.GetRequestStream()
   at OpenQA.Selenium.Remote.HttpCommandExecutor.Execute(Command commandToExecute)
   at OpenQA.Selenium.Remote.DriverServiceCommandExecutor.Execute(Command commandToExecute)
   at OpenQA.Selenium.Remote.RemoteWebDriver.Execute(String driverCommandToExecute, Dictionary`2 parameters)
TearDown : OpenQA.Selenium.WebDriverException : A exception with a null response was thrown sending an HTTP request to the remote WebDriver server for URL http://localhost:54785/session//window. The status of the exception was ConnectFailure, and the message was: Unable to connect to the remote server
  ----> System.Net.WebException : Unable to connect to the remote server
  ----> System.Net.Sockets.SocketException : No connection could be made because the target machine actively refused it 127.0.0.1:54785

------- End Stack Trace ------

If they are running interleaved, they shouldn't be. I thought NUnit took care of that so that they wouldn't step on each other.

**Comment by harishdhage** [comment]
I am also facing this issue with FF, please find the details below

Firefox version  : 56.0 (32-bit)
Geko driver : geckodriver-v0.19.0-win64
Webdriver version : selenium-server-standalone-3.6.0

code as below

package openBrow;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.testng.annotations.Optional;
import org.testng.annotations.Parameters;
import org.testng.annotations.Test;

public class Launch {

	//public static void main(String[] args) {
		// TODO Auto-generated method stub

		@Test
		public void launchBrow(){
			
		System.setProperty("webdriver.gecko.driver", "D:/Test Projects/ActiTime/src/drivers/geckodriver-v0.19.0-win64/geckodriver.exe");
		WebDriver driv = new FirefoxDriver();
		driv.get("http://localhost/login.do/");
		
		driv.findElement(By.id("username")).sendKeys("manager");
		driv.findElement(By.name("pwd")).sendKeys("admin");
		}
	}

Error log as below

[TestNG] Running:
  C:\Users\Harish\AppData\Local\Temp\testng-eclipse--915848218\testng-customsuite.xml

FAILED: launchBrow
org.openqa.selenium.WebDriverException: org.apache.http.conn.HttpHostConnectException: Connect to localhost:18452 [localhost/127.0.0.1, localhost/0:0:0:0:0:0:0:1] failed: Connection refused: connect
Build info: version: '3.6.0', revision: '6fbf3ec767', time: '2017-09-27T16:15:40.131Z'
System info: host: 'HARISH-CCS', ip: '172.20.10.9', os.name: 'Windows 7', os.arch: 'x86', os.version: '6.1', java.version: '1.8.0_92'
Driver info: driver.version: FirefoxDriver
	at org.openqa.selenium.remote.service.DriverCommandExecutor.execute(DriverCommandExecutor.java:92)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:586)
	at org.openqa.selenium.remote.RemoteWebDriver.startSession(RemoteWebDriver.java:217)
	at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:140)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:120)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:98)
	at openBrow.Launch.launchBrow(Launch.java:21)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at org.testng.TestRunner.privateRun(TestRunner.java:756)
	at org.testng.TestRunner.run(TestRunner.java:610)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:387)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:382)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
	at org.testng.SuiteRunner.run(SuiteRunner.java:289)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1293)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1218)
	at org.testng.TestNG.runSuites(TestNG.java:1133)
	at org.testng.TestNG.run(TestNG.java:1104)
	at org.testng.remote.AbstractRemoteTestNG.run(AbstractRemoteTestNG.java:132)
	at org.testng.remote.RemoteTestNG.initAndRun(RemoteTestNG.java:236)
	at org.testng.remote.RemoteTestNG.main(RemoteTestNG.java:81)
Caused by: org.apache.http.conn.HttpHostConnectException: Connect to localhost:18452 [localhost/127.0.0.1, localhost/0:0:0:0:0:0:0:1] failed: Connection refused: connect
	at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:159)
	at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:359)
	at org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:381)
	at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:237)
	at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:185)
	at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:89)
	at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:111)
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:185)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:72)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:56)
	at org.openqa.selenium.remote.internal.ApacheHttpClient.fallBackExecute(ApacheHttpClient.java:138)
	at org.openqa.selenium.remote.internal.ApacheHttpClient.execute(ApacheHttpClient.java:86)
	at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:337)
	at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:136)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:142)
	at org.openqa.selenium.remote.service.DriverCommandExecutor.execute(DriverCommandExecutor.java:83)
	... 31 more
Caused by: java.net.ConnectException: Connection refused: connect
	at java.net.DualStackPlainSocketImpl.waitForConnect(Native Method)
	at java.net.DualStackPlainSocketImpl.socketConnect(Unknown Source)
	at java.net.AbstractPlainSocketImpl.doConnect(Unknown Source)
	at java.net.AbstractPlainSocketImpl.connectToAddress(Unknown Source)
	at java.net.AbstractPlainSocketImpl.connect(Unknown Source)
	at java.net.PlainSocketImpl.connect(Unknown Source)
	at java.net.SocksSocketImpl.connect(Unknown Source)
	at java.net.Socket.connect(Unknown Source)
	at org.apache.http.conn.socket.PlainConnectionSocketFactory.connectSocket(PlainConnectionSocketFactory.java:75)
	at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:142)
	... 46 more


===============================================
    Default test
    Tests run: 1, Failures: 1, Skips: 0
===============================================


===============================================
Default suite
Total tests run: 1, Failures: 1, Skips: 0
===============================================

[TestNG] Time taken by org.testng.reporters.SuiteHTMLReporter@14751b3: 108 ms
[TestNG] Time taken by org.testng.reporters.EmailableReporter2@d2d9cb: 9 ms
[TestNG] Time taken by org.testng.reporters.jq.Main@1909a3: 163 ms
[TestNG] Time taken by [FailedReporter passed=0 failed=0 skipped=0]: 10 ms
[TestNG] Time taken by org.testng.reporters.XMLReporter@19b00e0: 8 ms
[TestNG] Time taken by org.testng.reporters.JUnitReportReporter@12e5a8e: 6 ms

**Comment by rgadeus** [comment]
Just wondering this issue is resolved for anyone, Appreciate your help - I am also getting same error in one of my machines - however same code and same Firefox browser is working in other machine

Selenium 3.4
Firefox 56.1
Gecko driver 0.19

org.apache.http.conn.HttpHostConnectException: Connect to localhost:4124 [localhost/127.0.0.1, localhost/0:0:0:0:0:0:0:1] failed: Connection refused: connect
Build info: version: 'unknown', revision: 'unknown', time: 'unknown'
System info: host: 'IBMT450PC062YD0', ip: '10.26.87.45', os.name: 'Windows 7', os.arch: 'x86', os.version: '6.1', java.version: '1.8.0_60'
Driver info: driver.version: Project_Master_Driverorg.apache.http.conn.HttpHostConnectException: Connect to localhost:4124 [localhost/127.0.0.1, localhost/0:0:0:0:0:0:0:1] failed: Connection refused: connect
Build info: version: 'unknown', revision: 'unknown', time: 'unknown'

**Comment by harishdhage** [comment]
This issue is resolved once I downgraded FF, Selenium Standalone driver.

I figured out the issue is because Selenium Standalone Server 3 is only supported in 64 bit Windows.

**Comment by ajsnyde** [comment]
I'm hitting similar issue, running newest release of 64-bit firefox and geckodriver on 64-bit Windows 10.

**Comment by AnujDhall** [comment]
Getting below error while trying to invoke IE browser

IE Explorer is 11.0
IE Driver is 3.11
Sel version is  3.11
please help , what can be the cause.
<<
Started InternetExplorerDriver server (64-bit)
3.11.1.0
Listening on port 3159
Only local connections are allowed
Exception in thread "main" org.openqa.selenium.WebDriverException: java.net.ConnectException: Failed to connect to localhost/0:0:0:0:0:0:0:1:3159
Build info: version: '3.11.0', revision: 'e59cfb3', time: '2018-03-11T20:33:08.638Z'
System info: host: 'INDA203920', ip: '10.164.47.61', os.name: 'Windows 7', os.arch: 'amd64', os.version: '6.1', java.version: '1.8.0_161'
Driver info: driver.version: InternetExplorerDriver
	at org.openqa.selenium.remote.service.DriverCommandExecutor.execute(DriverCommandExecutor.java:92)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:545)
>>

**Comment by rogerdc** [comment]
For me, it was a programming error. I had a BrowserFactory class that maintained a "driver" object across test executions. I made sure that when I tore down the browser, I nulled out the driver object as well.

**Comment by harishmuralidharan25** [comment]
I have seen the same error for my simple setup using selenium 3 and firefox 76. The funny part is if i pass the correct path to set system property
System.setProperty("webdriver.gecko.driver", "E:\\Software\\geckodriver-v0.24.0-win64\\geckodriver.exe");
WebDriver driver = new FirefoxDriver();
		driver.get("http://www.google.com");

The code works fine and the browser launches.
If i keep the geekodriver attached in project and give path as 
System.setProperty("webdriver.gecko.driver", "src/geeko/geckodriver.exe");
WebDriver driver = new FirefoxDriver();
		driver.get("http://www.google.com");
It gives me error,
Exception in thread "main" org.openqa.selenium.WebDriverException: java.net.ConnectException: Failed to connect to localhost/0:0:0:0:0:0:0:1:34477
Build info: version: '3.141.59', revision: 'e82be7d358', time: '2018-11-14T08:25:53'
System info: host: 'DESKTOP-7P20CGQ', ip: '192.168.1.205', os.name: 'Windows 10', os.arch: 'amd64', os.version: '10.0', java.version: '1.8.0_71'
Driver info: driver.version: FirefoxDriver
	at org.openqa.selenium.remote.service.DriverCommandExecutor.execute(DriverCommandExecutor.java:92)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:552)
	at org.openqa.selenium.remote.RemoteWebDriver.startSession(RemoteWebDriver.java:213)
	at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:131)

I dont have clue as to why. It is able to access the same driver there.

**Comment by diemol** [comment]
This thread has diluted from its original topic, I am going to close this issue since no more information from the original post has been provided. In case the issue is still present with the latest releases, please feel free to open a new issue with the complete template fiels filled out.
